### PR TITLE
perf: use an internal module resolver to improve require performance

### PR DESCRIPTION
**Pull request by MarshallOfSound**
### Context

I was doing some performance analysis why a bundled application with babel / webpack actually performed slower in some cases then a non-bundled application.  This PR is the result of those findings which I will document below.

#### Differences between `babel` and non-babel

This is a base example of a common JS file that uses `Menu`. It destructures from the require and then re-uses the `Menu` variable three times.  Note how the getter for `electron.Menu` is only accessed once and the result re-used, this will be important later.

```js
// Raw JS
const { Menu } = require('electron');

console.log(Menu);
console.info(Menu);

doSomethingWith(Menu);
```

This is an example using fancy ES6 imports and babel.

```js
import { Menu } from 'electron';

console.log(Menu);
console.info(Menu);

doSomethingWith(Menu);
```

At first glance this would result in the same thing (a destructure and then re-use of the `Menu` variable, but that is not the case.  The transpiled output of `babel` is as follows.

```js
'use strict';

var _electron = require('electron');

console.log(_electron.Menu);
console.info(_electron.Menu);

doSomethingWith(_electron.Menu);
```

Notice how in the transpiled case the `electron.Menu` getter is actually called three times instead of re-using the variable.  This is the core difference I found, below is an explanation on why this difference impacts performance in a negative way.

#### Why is accessing `electron.*` getters multiple times a bad thing?

If you look at how we actually declare the exports for `require('electron')` we map each export to property with a "getter" that returns the result of a `require` call.

https://github.com/electron/electron/blob/master/lib/browser/api/exports/electron.js#L9-L12

`require` calls using relative paths are actually terrible for performance especially on Windows, on macOS each require call with a relative path can take anywhere between 1 and 5ms throughout my testing. (5ms being the outlier, it averages between 1 and 2).  On windows those numbers would be even worse.

What this means for our `electron.Menu` getter is that every time we just **use** `electron.Menu` what we throught is a re-usable import variable is actually calling a heavy syncronous require call . **every single time**.  In some files you can use `Menu` or `MenuItem` or `app` tens if not hundreds of times for complex apps with big feature sets.  This results in an `O(n)` relationship between uses of `electron.*` modules and `require` calls.

The proposed (and implemented) solutions to these performance issues are explained below and map to the first and second commmits in this PR respectively.

### Solution #1 - Use an internal module resolver

Currently the initial `require('electron')` call is dramatically faster than a relative require, this is because of how nodes module resolution logic works (slowly).  We shortcut the path resolution for `require('electron')` by [forcefully returning the absolute path](https://github.com/electron/electron/blob/master/lib/common/reset-search-paths.js#L29-L37) that we know exists.

This solution involves doing a similar thing, but for all of our internal requires, not just `electron`.  It introduces a new **magic** module called `@electron/internal` that we map to require files inside our `electron.asar`.  E.g. `browser/init.js` becomes `@electron/internal/browser/init.js` and the results speak for themselves

#### Benchmarks #1

Timing the **JS only** initialization process of Electron (excluding the default app) we get the following nunbers averaged over 10 runs.

* Before this PR: `main-init: 92.158ms`
* Afte this PRr: `main-init: 59.130ms`

That's a 30% performance improvement on our intiailization logic.

### Solution #2 - Memoize the export getters of the `electron` repo

Even though Solution #1 had a noticeable performance improvement we are still going through the entire node `require` method every time someone re-uses an electron export (which is still slow even if we shortcut the path resolution).

Because node caches every module in a gloabl module cache anyway, we can cache our electron exports every time a user requests them so that regardless of how many times they are used and regardless of how many files use them we only load them once.  Again the resulting benchmarks speak for themselves.

#### Benchmarks #2

Given the following benchmark code running in the main process

```js
console.time('benchmark')
for (let i = 0; i < 1000; i += 1) {
  global.foo = electron.Menu
}
console.timeEnd('benchmark')
```

We got the following benchmarks for before / after our memoization optimization.

* Before this PR: `benchmark: 22.568.ms`
* After this PR: `benchmark: 0.241ms`

TLDR: Things are faster now 😆 

Notes: Performance improvements during startup and when requiring the "electron" module

### CODE REVIEW

**Comment by ckerr** [comment]
In addition to this PR, you should also report this upstream to Babel. They'd likely appreciate the suggestion on how to improve performance to all their users.

**Comment by ckerr** [comment]
This is pretty cool!

Is the internal resolver needed? It looks like the memoizer would give direct improvements for every call, while the resolver would only give improvements for the first call. The benchmarks seem to back this up: the resolver gives a 30% improvement, but the memoizer is more like two orders of magnitude better.

**Comment by MarshallOfSound** [comment]
> Is the internal resolver needed?

The internal resolver helps with our internal require calls so all of our files, things like `error-utils` and `rpc-server` and all those files that would otherwise be required with relative paths.  It's a one-off perf improvement that isn't as good as the memoize stuff, but it's still noticeable enough to be included.  You're right though, the memoize stuff is many times more helpful 😄 Dropping init time by ~30ms is also pretty nice 😆 #smallsteps

File: lib/browser/api/exports/electron.js [hunk 1]
```diff
@@ -1,13 +1,13 @@
-const common = require('../../../common/api/exports/electron')
+const common = require('@electron/internal/common/api/exports/electron')
```
Comment (jkleinsc) [inline lib/browser/api/exports/electron.js]: This is just a side effect, but I like the absolute pathing this introduces vs the relative pathing we were using.
